omegaconf/_utils.py

```diff
@@ -765,3 +730,58 @@ def is_container_annotation(type_: Any) -> bool:
 
 def is_generic_container(type_: Any) -> bool:
     return is_generic_dict(type_) or is_generic_list(type_)
+
+
+def _make_hashable(x: Any) -> Any:
+    """
+    Obtain a hashable version of `x`.
+
+    This is achieved by turning into tuples the lists and dicts that may be
+    stored within `x`.
+    Note that dicts are sorted, so that two dicts ordered differently will
+    lead to the same resulting hashable key.
+
+    :return: a hashable version of `x` (which may be `x` itself if already hashable).
+    """
+    # Hopefully it is already hashable and we have nothing to do!
+    try:
+        hash(x)
```

Comment:
Well, right now OmegaConf only supports a limited number of types, so I doubt this can be an issue.

Also, if it's expensive we're going to pay the price at some point anyway, since the cache is a dictionary and thus the hash function will be called when storing the data in the cache.

That being said, I think there are edge case where with my implementation of `_make_hashable()` we would end up hashing many times the same object, which could slow things down if the hash is expensive. The only way I can see around that would be to always inspect the content of `x` (going deep down containers), but that would slow down the more "standard" use cases so it doesn't seem like a good trade-off to me.

Reply:
> I feel like we did discuss this option briefly but I am not sure if the reasons that led us to the current decision are the right ones.

IIRC in my original version I had disabled the cache by default, but you preferred to keep it.
Given the advanced state of this PR, I'd rather keep it that way here, and then we can discuss potential changes before 2.1 is released.

For what it's worth, personally I think caching should also depend on the node, for instance if I have:
```
a: ${random:}
b: ${random:}
```
then I probably want `a` and `b` to be different, but each to keep the same value throughout program execution.

Reply:
As discussed in another comment, I suggest that we continue this discussion in #526 and resolve it here.